### PR TITLE
Geometry_Engine: Add join mesh method and fix MergeVertices

### LINKED ISSUE #2945: Geometry_Engine: Add method for joining Meshes into a single Mesh.
#### Description:

Method to be added for Joining multiple Meshes to a single Mesh.

Can add toggles for how complex the Joining is to be made, with toggle if the Vertices should be merged or not.

### LINKED ISSUE #2946: Geometry_Engine: MergedVertices is not functioning
#### Description:

Method not working. Is not returning a mesh with culled vertices and correct faces:

Before:
image

After:

image

#### Steps to reproduce:

Use the method with a mesh with duplicate vertices

#### Expected behaviour:

For the method to return a valid mesh with the same amount of faces but with duplicate vertices culled

### PR DESCRIPTION
**Pull request by IsakNaslundBh**
### Issues addressed by this PR

Closes #2945 
Closes #2946 

- Adding method for joining multiple meshes into single mesh. Allows for (does not check for) if the meshes are disjointed or not.
- Fix the non-functioning MergedVertices method and rename it MergeVertices (removing the d). Complete rewrite as the previous version was not functioning, and we now have access to better grouping functionality through the domain tree clusters.

### Test files

https://burohappold.sharepoint.com/:f:/s/BHoM/Eo4Nblb9345DlWdrFSNZwx8BwNsgwKB74BxCuqah5qCYPg?e=Hmavj2

### CODE REVIEW

**Comment by IsakNaslundBh** [comment]
@BHoMBot check compliance
@BHoMBot check unit-tests

File: Geometry_Engine/Modify/MergeVertices.cs [hunk 1]
```diff
@@ -34,104 +39,78 @@ public static partial class Modify
...
-                vertices[face.B].Faces.Add(face);
-                vertices[face.C].Faces.Add(face);
+            if (mesh == null)
+                return null;
 
-                if (face.IsQuad())
-                    vertices[face.A].Faces.Add(face);
-            }
+            //Set up list on structs containing location and index
+            List<Tuple<Point,int>> vertices = mesh.Vertices.Select((x, i) => new Tuple<Point, int>(x, i)).ToList();
```
Comment (pawelbaran) [inline Geometry_Engine/Modify/MergeVertices.cs:53]: Just curious, cant we switch to tuples written as `(Point, int)`?

File: Geometry_Engine/Compute/Join.cs [hunk 2]
```diff
@@ -204,6 +207,52 @@ public static List<PolyCurve> IJoin(this List<ICurve> curves, double tolerance =
...
+        /***************************************************/
+        /****                Join meshes                ****/
+        /***************************************************/
+
+        [Description("Joins a list of Meshes into a single Mesh. Disjointed Meshes are allowed to be joined into one mesh.")]
+        [Input("meshes", "The meshes to join.")]
+        [Input("mergeVertices", "If true, duplicate vertices will be merged. If false, duplicate vertices will be kept.")]
+        [Input("tolerance", "Only used if mergeVertices is true. The maximum allowable distance between two vertices for them to be deemed the same vertex.", typeof(Length))]
+        [Output("mesh", "The joined meshes as a single mesh.")]
+        public static Mesh Join(this List<Mesh> meshes, bool mergeVertices = false, double tolerance = Tolerance.Distance)
```
[the remark this answers is not in the document]
Comment (IsakNaslundBh) [reply Geometry_Engine/Compute/Join.cs:219]: would very much disagree that it is useless without merging vertices. combining two structures into one could still be of use.

Happy to change default to true though.

File: Geometry_Engine/Compute/Join.cs [hunk 3]
```diff
@@ -204,6 +207,52 @@ public static List<PolyCurve> IJoin(this List<ICurve> curves, double tolerance =
...
+                Vertices = meshes[0].Vertices.ToList(),
+                Faces = meshes[0].Faces.ToList()
+            };
+
+            for (int i = 1; i < meshes.Count; i++)  //Add on the rest
+            {
+                Mesh mesh = meshes[i];
+                int vertexCount = returnMesh.Vertices.Count;
+
+                returnMesh.Vertices.AddRange(mesh.Vertices);
```
[the remark this answers is not in the document]
Comment (IsakNaslundBh) [reply Geometry_Engine/Compute/Join.cs:251]: good question. the way I change MergeVertices to now utilises the DomainTreesClustering which I think is quite efficient, and requires you to send in all the data in one go.

Might be even more efficient ways to handle it here though, but think what is in this PR is efficient enough, and makes it possible to keep this method simpler, without having to introduce complex structures like domain trees/point matrix in here.

File: Geometry_Engine/Modify/MergeVertices.cs (hunk 1 above)
Comment (IsakNaslundBh) [reply Geometry_Engine/Modify/MergeVertices.cs:53]: possibly. Not sure what version of c# that requires though, compared to which version we officially support.

**Comment by IsakNaslundBh** [comment]
@BHoMBot check compliance

File: Geometry_Engine/Compute/Join.cs (hunk 2 above)
[the remark this answers is not in the document]
Comment (pawelbaran) [reply Geometry_Engine/Compute/Join.cs]: Ha, I meant _virtually useless to me_ 😉 likely to be my ignorance. Regardless, I think merging vertices makes more sense as a default, glad you do not mind the change 👍

**Comment by pawelbaran** [comment]
@BHoMBot check core
@BHoMBot check unit-tests
@BHoMBot check compliance

**Review by pawelbaran — APPROVED** [review]
Approving based on the code review, answered comments as well as testing made using provided test file as well as in promptu performance tests, which showed there is still space for improvement in finding duplicate points (but this refers to BHoM cull duplicate methods in general).

**Pull request merged**